### DISCUSSION TITLE
Unable to export log to otel using log4j2 appender

### DISCUSSION THREAD

**Discussion by mohit32573** [open-telemetry/opentelemetry-java-instrumentation#12420] [Q&A]
I am trying to setup up log export to otel collector, but unable to complete. I am using log4j2 appender. But getting different error. I am not using spring or spring boot.  
Thanks in advance

![17284679773365663663721030477893](https://github.com/user-attachments/assets/646793b3-6350-43d2-9243-fdec2b9d5f0b)
![17284680318988135258032610170088](https://github.com/user-attachments/assets/1f3035fc-57af-4100-8516-d164b7f2fa43)

**Comment by jeanbisutti** [comment]
You can refer to this example: https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/log4j/log4j-appender-2.17/library/README.md#usage

**Reply by mohit32573** [reply]
@jeanbisutti  Thanks for the reply, the error got resolved. But still i am not able to see any logs on OTEL Collector. Just want to know where i can put the endpoint , servicename, log level etc. when I am adding the attributes in log4j2 configuration under otelappender , it is giving error for invalid attributes

**Comment by jeanbisutti** [comment]
@mohit32573 The available settings are described here: https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/log4j/log4j-appender-2.17/library/README.md#settings-for-the-log4j-appender

Have you followed this documentation?

**Reply by mohit32573** [reply]
@jeanbisutti  - Yes i have already tried. the settings for exporter seems not available.

**Comment by mohit32573** [comment]
Can any one help on this ? Thanks

**Comment by jeanbisutti** [comment]
@mohit32573 Your configuration in the second screenshot is different from the configuration described in the doc: `<OpenTelemetry`, https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/log4j/log4j-appender-2.17/library/README.md#settings-for-the-log4j-appender

Have you updated your configuration? If it still does not work, please provide your last configuration.

**Reply by mohit32573** [reply]
Yes the configuration is updated, but still getting error
![New Configuration](https://github.com/user-attachments/assets/1a2c080f-feaf-4ad0-9790-b4a9d4a2c86d)
![new error](https://github.com/user-attachments/assets/b8195271-f493-4ebf-890d-00b3dfdd8a94)

**Comment by jeanbisutti** [comment]
Have you added [this dependency](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/log4j/log4j-appender-2.17/library/README.md#add-these-dependencies-to-your-project)?

**Reply by mohit32573** [reply]
Yes
![image](https://github.com/user-attachments/assets/1cadd5c2-de46-4ea4-aa29-f0f21d0cfd0d)

**Comment by jeanbisutti** [comment]
I have tried to reproduce the issue from the information you have provided, but it seems to work on my side. Could you please provide a project reproducing the problem?

**Reply by mohit32573** [reply]
The error I am getting is this 
![image](https://github.com/user-attachments/assets/7f6c3748-5045-4250-99f2-bc288c953ef1)

Runtime Arguments I am using :

![runtimeargs](https://github.com/user-attachments/assets/db5d8864-ae04-417b-8310-147bae5d4efc)

Dependency Version : 
![image](https://github.com/user-attachments/assets/c780ca5b-36bc-4241-8609-e300b40e6881)

Can you please let me know what all other dependencies are required. Also if you can help me the code snippet used to instantiate the OpenTelemetrySdk.  I am talking about this : 

![image](https://github.com/user-attachments/assets/fd6a8779-7403-4b76-92b1-554db7a4288a)

**Reply by jeanbisutti** [reply]
Could you please provide a Github repository reproducing the issue?

**Reply by mwagner5310** [reply]
@jeanbisutti is it necessary to manually include the dependency via Maven like descriped in the documentation here:
https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/log4j/log4j-appender-2.17/library/README.md#add-these-dependencies-to-your-project

I am asking because in the following tutorial they said, that the "log4j log appender is bundled with the java agent and will be installed automatically"
https://docs.honeycomb.io/send-data/logs/opentelemetry/sdk/java/

I am using the OTEL Kubernetes Operator with Auto-Instrumentation my JAVA Application.
When I don't manually add the appender as Dependency in Maven, I get the same Error as mohit32573.

**Comment by jeanbisutti** [comment]
@mwagner5310 With the OpenTelemetry Java agent, you don't need to include https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/log4j/log4j-appender-2.17/library/README.md#add-these-dependencies-to-your-project to have the logs instrumented.

**Reply by mwagner5310** [reply]
Thanks for your fast reply.

But do you have any ideas why I am still getting the following Errors indicating that the dependency is missing.
```
2025-04-02T08:40:19.000+0200 Initializing Log4J
2025-04-02T08:40:19.000+0200 2025-04-02T08:40:41.592462388Z main ERROR Appenders contains an invalid element or attribute "OpenTelemetry"
2025-04-02T10:57:19.899+0200 2025-04-02T08:40:41.592462388Z main ERROR Appenders contains an invalid element or attribute "OpenTelemetry"
2025-04-02T10:57:19.899+0200 2025-04-02T08:40:41.632674456Z main ERROR Unable to locate appender "OpenTelemetryAppender" for logger config "root"   
```

Do I need to configure the Java Agent to include the Log4j Appender?

**Reply by jeanbisutti** [reply]
It seems that you have added the following kind of code in the log4j xml file:

```
<Appenders>
  <OpenTelemetry name="OpenTelemetryAppender"
  />
</Appenders>
```

It's not needed if you use the OTel Java agent.

**Reply by mwagner5310** [reply]
Oh yes, you are right.
I still get the logs after removing the OTEL Appenders and also the Loggers.

That is even better for me, so that I must not change the log4j config of existing projects.
Thanks for your help.
Maybe this can be added to the documentation for other users which running into the same problem.
